pywhy_stats/conditional_ksample/kcd.py

```diff
@@ -0,0 +1,262 @@
+        kernel for categorical data. Note that we currently only consider string values as categorical data.
+        Kernels can be specified in the same way as for :func:`~sklearn.metrics.pairwise.pairwise_kernels`
+        with the addition that 'delta' kernel is supported for categorical data.
+    null_sample_size : int
+        The number of samples to generate for the bootstrap distribution to approximate the pvalue,
+        by default 1000.
+    normalize_data : bool
+        Whether the data should be standardized to unit variance, by default True.
+    propensity_model : Optional[sklearn.base.BaseEstimator], optional
+        The propensity model to use to estimate the propensity score, by default None.
+    propensity_weights : Optional[ArrayLike], optional
+        The propensity weights to use, by default None, which means that the propensity scores
+        will be estimated from the propensity_model.
+    centered : bool
+        Whether the kernel matrix should be centered, by default True.
+    n_jobs : Optional[int], optional
+        The number of jobs to run in parallel, by default None.
+    random_seed : Optional[int], optional
+        Random seed, by default None.
+
+    Notes
+    -----
+    Any callable can be given to create the kernel matrix. For instance, to use a particular kernel
+    from sklearn::
+
+        kernel_X = func:`sklearn.metrics.pairwise.pairwise_kernels.polynomial`
+
+    In addition, we implement an efficient delta kernel. The delta kernel can be specified using the
+    'kernel' string argument.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    test_statistic, pvalue = _kernel_test(
+        X=X,
+        Y=Y,
+        group_ind=group_ind,
+        kernel_X=kernel_X,
+        kernel_Y=kernel_Y,
+        propensity_weights=propensity_weights,
+        propensity_model=propensity_model,
+        null_sample_size=null_sample_size,
+        normalize_data=normalize_data,
+        centered=centered,
+        n_jobs=n_jobs,
+        random_seed=random_seed,
+    )
+    return PValueResult(pvalue=pvalue, statistic=test_statistic)
+
+
+def _kernel_test(
+    X: ArrayLike,
+    group_ind: ArrayLike,
+    Y: ArrayLike,
+    kernel_X: Optional[Callable[[np.ndarray], np.ndarray]],
+    kernel_Y: Optional[Callable[[np.ndarray], np.ndarray]],
+    propensity_weights: Optional[ArrayLike],
+    propensity_model: Optional[BaseEstimator],
+    null_sample_size: int,
+    normalize_data: bool,
+    centered: bool,
+    n_jobs: Optional[int],
+    random_seed: Optional[int],
+) -> Tuple[float, float]:
+    X, Y, _ = _preprocess_kernel_data(X, Y, normalize_data=normalize_data)
+    _preprocess_propensity_data(
+        group_ind=group_ind,
+        propensity_weights=propensity_weights,
+        propensity_model=propensity_model,
+    )
+
+    enc = LabelBinarizer(neg_label=0, pos_label=1)
+    group_ind = enc.fit_transform(group_ind)
```

Comment:
For my understanding, we only need this line in case the group IDs are not 0 and 1, right?

Reply:
Yes this is mainly to simplify the internal API later.